clippy_lints/src/duration_suboptimal_units.rs

```diff
@@ -0,0 +1,181 @@
+            //
+            // let dur = Duration::from_secs(SIXTY);
+            //           ^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try: `Duration::from_mins(1)`
+            let Some(Constant::Int(value)) = ConstEvalCtxt::new(cx).eval_local(arg, arg.span.ctxt()) else {
+                return;
+            };
+            // There is no need to promote e.g. 0 seconds to 0 hours
+            if value == 0 {
+                return;
+            }
+
+            let function_name = func_name.ident.name.as_str();
+            let Some((promoted_function_name, promoted_value)) = self.promote(cx, function_name, value) else {
+                return;
+            };
+
+            span_lint_and_sugg(
+                cx,
+                DURATION_SUBOPTIMAL_UNITS,
+                expr.span,
+                "constructing a `Duration` using a smaller unit when a larger unit would be more readable",
+                "try",
+                format!("Duration::{promoted_function_name}({promoted_value})"),
+                Applicability::MachineApplicable,
+            );
+        }
+    }
+}
+
+impl DurationSuboptimalUnits {
+    /// Tries to promote the given function and value to a bigger time unit and returns the promoted
+    /// function name and value.
+    ///
+    /// Returns [`None`] in case no promotion could be done.
+    fn promote<'a>(&self, cx: &LateContext<'_>, function_name: &'a str, value: u128) -> Option<(&'a str, u128)> {
+        let mut promoted = false;
+        let mut promoted_function_name = function_name;
+        let mut promoted_value = value;
+
+        let bigger_units = UNITS
+            .iter()
+            .skip_while(|unit| unit.function_name != function_name)
+            .skip(1);
+
+        for bigger_unit in bigger_units {
+            if !self.msrv.meets(cx, bigger_unit.const_support_added) {
+                // We have to break early, as we can not skip versions.
+                // We can't skip any, as they are needed to correctly calculate the promoted value.
+                break;
+            }
+
+            if promoted_value.is_multiple_of(bigger_unit.factor) {
+                promoted = true;
+                promoted_value /= bigger_unit.factor;
+                promoted_function_name = bigger_unit.function_name;
+            } else {
+                break;
+            }
+        }
+
+        promoted.then_some((promoted_function_name, promoted_value))
+    }
+}
+
+struct Unit {
+    /// Name of the function on [`Duration`](std::time::Duration) to construct it from the given
+    /// unit, e.g. [`Duration::from_secs`](std::time::Duration::from_secs)
+    function_name: &'static str,
+
+    /// The increase factor over the previous (smaller) unit
+    factor: u128,
+
+    /// In what rustc version support for this function in const contexts was added
+    const_support_added: RustcVersion,
+}
+
+const UNITS: &[Unit] = &[
+    Unit {
+        function_name: "from_nanos",
```

Comment:
```suggestion
        function_name: sym::from_nanos,
```

(and similarly for the others)